File: app/src/main/java/com/witype/Dragger/mvp/BaseActivity.java
package com.witype.Dragger.mvp;

import android.app.Activity;
import android.os.Bundle;

import com.trello.rxlifecycle3.components.support.RxAppCompatActivity;
import com.witype.Dragger.app.MApplication;
import com.witype.Dragger.di.component.AppComponent;
import com.witype.Dragger.mvp.contract.IBaseView;
import com.witype.Dragger.mvp.present.IBasePresenter;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public abstract class BaseActivity<P extends IBasePresenter> extends RxAppCompatActivity implements IBaseView {

    @Inject
    P presenter;

    private Unbinder bind;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setupComponent(((MApplication) getApplication()).getAppComponent());
        setContentView(getResId());
        initView();
        if (presenter != null) {
            presenter.onStart();
        }
        initData();
    }

    public abstract int getResId();

    public void initView() {
        bind = ButterKnife.bind(this);
    }

    protected void initData() {

    }

    @Nullable
    public P getPresenter() {
        return presenter;
    }

    protected abstract void setupComponent(@NonNull AppComponent appComponent);

    @Override
    public void showToast(String message) {

    }

    @Override
    public void showLoading(String message) {

    }

    @Override
    public void showLoading() {

    }

    @Override
    public void dismissLoading() {

    }

    @Override
    public void showToast(int resId) {
        showToast(getString(resId));
    }

    @Override
    public Activity getActivity() {
        return this;
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (presenter != null) {
            presenter.onStop();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (presenter != null) {
            presenter.onDestroy();
        }
        if (bind != null) {
            bind.unbind();
        }
    }
}

File: kotlindemo/src/main/java/com/witype/kotlindemo/mvp/ui/BaseActivity.kt
package com.witype.kotlindemo.mvp.ui

import android.app.Activity
import android.os.Bundle
import android.widget.Toast
import butterknife.ButterKnife
import butterknife.Unbinder
import com.trello.rxlifecycle3.components.support.RxAppCompatActivity
import com.witype.kotlindemo.app.KotlinDemoApplication
import com.witype.kotlindemo.mvp.contract.IBaseView
import com.witype.kotlindemo.mvp.presenter.IBasePresenter
import com.witype.mvp.di.component.AppComponent
import javax.inject.Inject

abstract class BaseActivity<P : IBasePresenter> : RxAppCompatActivity(), IBaseView {

    @set:Inject
    internal var presenter: P? = null

    var bind: Unbinder? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupComponent((application as KotlinDemoApplication).getAppComponent())
        setContentView(resId)
        initView()
        if (presenter != null) {
            presenter?.onStart()
        }
        initData()
    }

    abstract val resId: Int

    open fun initView() {
        bind = ButterKnife.bind(this)
    }

    protected fun initData() {}

    protected abstract fun setupComponent(appComponent: AppComponent)

    override fun showToast(message: String) {
        Toast.makeText(this,message,Toast.LENGTH_LONG).show()
    }

    override fun showLoading(message: String) {}

    override fun showLoading() {}

    override fun dismissLoading() {}

    override fun showToast(resId: Int) {
        showToast(getString(resId))
    }

    override val activity: Activity
        get() = this

    override fun onStop() {
        super.onStop()
        presenter?.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter?.onDestroy()
        bind?.unbind()
    }
}